Allow collections to add RST files

felixfontein opened the pull request:
The RST files have to be in `docs/docsite/rst/` (and subdirs), need to use labels of the form `ansible_collections.{namespace}.{name}.docsite.` (RST files not satisfying that are ignored), and there needs to be a `index.yml` which allows to add sections to the main collection page and allows to say which files show up in the toctree for that section (also allows to directly reference things).

Example branches:
- https://github.com/ansible-collections/community.docker/compare/main...felixfontein:scenario-guide
- https://github.com/felixfontein/ansible-tools/compare/docs

These result in:
- https://ansible.fontein.de/collections/community/docker/ (note the scenario guide!)
- https://ansible.fontein.de/collections/felixfontein/tools/ (note the first three sections on filters and tests, which are undocumentable plugin types and thus are usually not shown at all!)

There's a lint tool for collection owners (`antsibull-lint collection-docs /path/to/collection`), it checks RST labels and also runs `rstcheck`.

The PR also contains a commit which ensures that "empty" collections (which have no documentable plugins) are shown in the index, and will obtain docs. Example: https://ansible.fontein.de/collections/felixfontein/acme/ (that collection only has roles; ~~I'll add docsite RST files for these eventually, currently I only have MD files~~ it now has role docs as well)

briantist commented:
I don't feel qualified to actually review this but it looks great. I could definitely see myself using this for c.hv!

felixfontein commented:
(I've just rebased so this PR includes all other docs building PRs that were merged today. Makes it easier to test for me :) )

felixfontein commented:
As discussed at the docs meeting, I've renamed `index.yml` to `extra-docs.yml`, and removed the currently not used entries code.

I've also updated the example https://github.com/ansible-collections/community.docker/compare/main...felixfontein:scenario-guide, which (still) results in https://ansible.fontein.de/collections/community/docker/.

ready_for_review

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,226 @@
+        except Exception as e:
+            errors.append(str(e))
+    elif report_not_existing:
+        errors.append('extra-docs.yml does not exist')
```
I think I'd change this to raise an exception if the rile does not exist and let the calling code decide whether to ignore the error or not.  That simplifies this code in two ways:

* This function does not need to know what the caller intends to do with the results and doesn't need to deviate depending on the callers intentions.
* This function can exit early when index_path is not a file, so we can drop a nesting level.

felixfontein replied:
Implemented in 4de665a7a0c51d9ec106adde46ef49f5e7f21e06.

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,226 @@
+    return sections, documents
+
+
+async def load_collections_docsites(collection_metadata: t.Mapping[str, AnsibleCollectionMetadata]
```
I notice that we're only using `collection_name` and `collection_metadata.path` from this data structure in this function.  In my vision for the code base, I'd like there to be a split between "application" level code and "library" level code.

Functions in the library code should try to do specific things with the minimum information necessary to do its job.

Functions in the application code can do more aggregated tasks with messy data being passed around.

I want to put all application code in cli or a subdirectory of cli (`cli/doc_commands`, for instance.).

I'm not sure precisely where the line should be drawn in this piece of code.  Maybe `load_collections_docsites()` should be application level code and live under cli/doc_commands somewhere and `load_collection_docsite()` would be library level code that takes the collection_name and path as input parameters?

felixfontein replied:
I've reduced the data passed to just the collection path in 4f222971f8e07fb7245e8a5b643633654ddcd5bc.

abadger commented:
I think we should use a name other than docsite for the names of these functions and variables.  We're already using docsite to refer to building docs.ansible.com as a whole.  How about one of the following:

* extra_docs
* manual_docs
* static_docs

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,234 @@
+        raise DocsiteIndexError('extra-docs.yml does not exist')
+
+    try:
+        index = load_yaml_file(index_path)
```
I note that this is synchronous code rather than async code.  The majority of time is going to be spent reading and writing the actual page data rather than the extra-docs.yml file so it's not the end of the world but it is a point where we could be doing things in parallel.  I assume you coded it as synchronous for easy reuse with antsibull-lint?

felixfontein replied:
The main reason is that the IO part of this function is done by `load_yaml_file`, which is not async either. It is also used to load routing data for collections, so eventually it should be converted to async, but once that's done adjusting the calling code is also not so hard, so I would defer adjusting this function until then. Is that ok for you?

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,234 @@
+                content = await f.read()
+
+            # Lint content
+            _, errors = lint_required_conditions(content, collection_name)
```
I generally try to avoid using ` bare `_` as an identifier since if we ever i18n'ize this program we'll want to use that for gettext.  In ansible-core they use `dummy` for variables that should be ignored.  Here's pylint's regex:
```
dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_
```

felixfontein replied:
I've renamed all occurences of `_` to `dummy`.

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,234 @@
+
+            # When no errors were found, add to output
+            if not errors:
+                documents.append((path_prefix + doc[1], content))
```
I'm a little worried about how much additional memory this might end up using but I'm not sure if I need to worry about that now or if it can be deferred until it actually becomes a problem.  I can see three strategies we could take:

* What you're doing.  We only read the files once.  We validate the data prior to further processing.
* Validate when we use the data (ie, when we copy it over to the docsite location).  This has the benefit of only having the data in memory when it is copied.  The drawback is that we don't know if the pages have valid data until we're ready to write it.
* Discard the data after validating and then read it again when we copy.  Benefits of both the above approaches.  Drawback is that we read the data twice which could be slower if the files aren't cached in memory by the OS.

We could also implement our own cache (something as simple as @functools.lru_cache() on a function which returns the contents of a full path to a file could work for small sets of extra_docs but it will probably have more cache misses than if we code something ourselves) to limit memory usage while also reading less than the full data set each time.

Do you think we should change this strategy now or wait until we start running into problems?

felixfontein replied:
> Discard the data after validating and then read it again when we copy.

I think that's the best approach. During a regular docs build, the chance that data changes between the two reads should be 0.

felixfontein replied:
Implemented in b927800e6798f18b6692e9d930009486096fdaf9.

abadger commented on antsibull/docsite_docs.py:
```diff
@@ -0,0 +1,234 @@
+    '''Load a docsite extra-docs.yml file.
+
+    :arg index_path: Path to extra-docs.yml (does not need to exist).
+    :arg report_not_existing: Whether to report when extra-docs.yml does not exist.
```
```suggestion
```

felixfontein commented:
> I think we should use a name other than docsite for the names of these functions and variables. We're already using docsite to refer to building docs.ansible.com as a whole. How about one of the following:
> 
>     * extra_docs

I renamed docsite to this in a3aa61dd88b5ad708375df17a9b3e89c7aaa34ab.

abadger commented:
Everything looks good!  Talked with felixfontein earlier and he will squash this manually so we can look at the saving-content-in-memory implementation later if we need to, then go ahead and merge it.

Thanks @felixfontein !

felixfontein commented:
@abadger @briantist thanks a lot for reviewing this!
@acozine @samccann @lmalivert thanks a lot for discussing this! :)
